Fraser/generator transformation graph

delicious-lemon opened the pull request:
This is a big PR.
- Everything has been split up into more manageable classes (instead of the one massive type family with everything in it).
- You can still do all the same stuff as before + there is a new feature - PermutingGenerator
  - this lets you write a parameterised generator as a strongly connected graph of model permuting generators that obey strict contracts
 - Backends are pluggable now - there is one for pure functions and one for Plutus Scripts
 - There is now a simple example describing both styles of generator  (you can use the old or new style)
   - these include Plutarch example tests
 - The README.md has been updated

delicious-lemon committed:
Remove dependency on Plutus and add Plutarch dependency

delicious-lemon committed:
Proper.Script

delicious-lemon committed:
model budget with bounds

delicious-lemon committed:
remove plutarch dependency

delicious-lemon committed:
remove constraint on base

delicious-lemon committed:
Revert "remove constraint on base"

This reverts commit 7ec7e70fd240ed56d0436f238e730a86a6bc7b10.

delicious-lemon committed:
Revert "remove plutarch dependency"

This reverts commit 4437625cd4756c6005e3f6779d69ff403eb0ae85.

delicious-lemon committed:
fix flake name

delicious-lemon committed:
a tic-tac-toe example specification

the goal was to make a non-trivial example specification
that generates more property tests than lines of code in the spec

the code really needs cleaned up and could be improved substantially
(I'm just checking this in as a draft for now)

TODO: cleanup + write some plutarch that meets the spec

delicious-lemon committed:
add some motivational and descriptive text to the readme

Full manual with examples coming soon.

delicious-lemon committed:
implemented Generator Transformation Graph and default generator

delicious-lemon committed:
implemented base generator for TicTacToe

-- most of the tests are failing due to limits rejection sampling

-- better idea for Transformations

```Haskell
class Transformation t p where
  transformation :: Formula p -- arrows mate

```

delicious-lemon committed:
set function based transforms

considering a different approach
  - could do a conflict directed search
    - first check if the set function reduces the conflict set
    - apply the model function
      - if a solution return
      - if not recurse

delicious-lemon committed:
hacky conflict directed search

delicious-lemon committed:
better transforms by matching on Formula based predicates

delicious-lemon committed:
log transformations on failure

delicious-lemon committed:
documentation and formatting

delicious-lemon committed:
better idea - define transformations that only change a single property

delicious-lemon committed:
this is better I think

delicious-lemon committed:
this seems usable

delicious-lemon committed:
remodelling TicTacToe some more

delicious-lemon committed:
using parameterised generator with transformations seems good

delicious-lemon committed:
transformation tree construction

delicious-lemon committed:
think the transformation invariant myst be a strict arrow

 encoding random multi-arrows results in a mess

 if we check for singleton solutions up front then we can build a graph

each edge on the graph will be a random function w.r.t model but a
deterministic function w.r.t. property sets

delicious-lemon committed:
building a graph first

delicious-lemon committed:
add test for strongly connected graph

delicious-lemon committed:
some refactoring

delicious-lemon committed:
example simple as possible spec

delicious-lemon committed:
remove a Proxy

delicious-lemon committed:
permuting generator

delicious-lemon committed:
simple example for permuting generator

delicious-lemon committed:
clean up some files

delicious-lemon committed:
plutus test runner

delicious-lemon committed:
add plutarch test examples

delicious-lemon committed:
update readme

delicious-lemon committed:
testing out permutation graph generator on tic-tac-toe

delicious-lemon committed:
note on overlapping edges - they are chosen randomly

delicious-lemon committed:
improved error message when graph not strongly connected

delicious-lemon committed:
add property ToBoardIsEmpty

delicious-lemon committed:
add property OddNumberOfPiecesOnBoard

delicious-lemon committed:
add prop equal number of pieces on board

delicious-lemon committed:
improve error output - model inconsistency information was missing

delicious-lemon committed:
change tagline

delicious-lemon committed:
new prop - OneMoreXThanO

delicious-lemon committed:
ideas for PermutingGenerator improvements

  - we can construct property tests for each edge automatically
    - we use bootstrap these tests using the PermutingGenerator
  - we can also check for redundant edges and suggest they are removed

delicious-lemon committed:
exhaustive model consistency test suite generator for PermutingGenerator

delicious-lemon committed:
fix errors in TicTacToe model revealed by new test suite generator

NOTE: deterministic path finding is an issue that needs resolved
we should find paths randomly instead since otherwise we may be hiding
errors introduced by Gen.filter which can crop up again when the graph
node numbering changes and and we end up taking different paths

delicious-lemon committed:
random path finding

delicious-lemon committed:
fix bug in graph distance algorithm

delicious-lemon committed:
add example usage for IntPermutingGen

delicious-lemon committed:
10x speedup on tests!

we are now randomly choosing between shortest paths using a distance
matrix. this is faster than doing depth first search to find a path for
each test. winning!

delicious-lemon committed:
rename property

delicious-lemon committed:
rename class Proposition to LogicalModel

delicious-lemon committed:
rename class HasProperties to HasLogicalModel

delicious-lemon committed:
rename PermutingGenerator to HasPermutationGenerator

delicious-lemon committed:
rename Device -> Pure

delicious-lemon committed:
some more renaming

delicious-lemon committed:
update readme

delicious-lemon committed:
rename run to script in HasPureTestRunner

delicious-lemon committed:
add TicTacToe property ToBoardEqualsFromBoard

delicious-lemon commented on examples/Spec/Int.hs:
```diff
@@ -0,0 +1,89 @@
+    | IsMinBound
+    deriving stock (Eq,Ord,Enum,Show,Bounded)
+
+instance LogicalModel IntProp where
```
Now we associate a logical model with the model properties directly.

delicious-lemon committed:
note on TicTacToe example + fallback pathgen for circuitous routes

delicious-lemon commented on examples/Spec/Int.hs:
```diff
@@ -0,0 +1,89 @@
+import Plutarch (compile)
+import Plutarch.Prelude
+
+data IntProp =
+      IsNegative
+    | IsPositive
+    | IsZero
+    | IsLarge
+    | IsSmall
+    | IsMaxBound
+    | IsMinBound
+    deriving stock (Eq,Ord,Enum,Show,Bounded)
+
+instance LogicalModel IntProp where
+  logic = ExactlyOne [Var IsNegative, Var IsPositive, Var IsZero]
+     :&&: ExactlyOne [Var IsLarge, Var IsSmall]
+     :&&: (Var IsZero :->: Var IsSmall)
+     :&&: (Var IsMaxBound :->: (Var IsLarge :&&: Var IsPositive))
+     :&&: (Var IsMinBound :->: (Var IsLarge :&&: Var IsNegative))
```
We can start simply with a set of Propositions and associated model logic. This logic defines which Propositions are valid in conjunction. For example `IsNegative` and `IsPositive` cannot both be true simultaneously.

delicious-lemon commented on examples/Spec/Int.hs:
```diff
@@ -0,0 +1,89 @@
+     :&&: (Var IsMaxBound :->: (Var IsLarge :&&: Var IsPositive))
+     :&&: (Var IsMinBound :->: (Var IsLarge :&&: Var IsNegative))
+
+instance HasLogicalModel Int IntProp where
+  satisfiesProperty i IsNegative = i < 0
+  satisfiesProperty i IsPositive = i > 0
+  satisfiesProperty i IsMaxBound = i == maxBound
+  satisfiesProperty i IsMinBound = i == minBound
+  satisfiesProperty i IsZero     = i == 0
+  satisfiesProperty i IsLarge    = i > 10 || i < -10
+  satisfiesProperty i IsSmall    = i <= 10 && i >= -10
```
From here we can associate a model datatype that the logical model applies to and define what the propositions mean in terms of this datatype.

delicious-lemon commented on examples/Spec/Int.hs:
```diff
@@ -0,0 +1,89 @@
+  satisfiesProperty i IsLarge    = i > 10 || i < -10
+  satisfiesProperty i IsSmall    = i <= 10 && i >= -10
+
+instance HasParameterisedGenerator Int IntProp where
+  parameterisedGenerator s = forAll $ do
+    i <- if IsZero `elem` s
+           then pure 0
+           else if IsSmall `elem` s
+                  then Gen.int (linear 1 10)
+                  else if IsMaxBound `elem` s
+                         then pure maxBound
+                         else Gen.int (linear 11 (maxBound -1))
+    if IsNegative `elem` s
+       then if IsMinBound `elem` s
+               then pure minBound
+               else pure (-i)
+       else pure i
```
Now we define a parameterisedGenerator that can generate a model that satisfies a set of properties given a set of properties.

delicious-lemon commented on examples/Spec/Int.hs:
```diff
@@ -0,0 +1,89 @@
+               else pure (-i)
+       else pure i
+
+intGenTests :: TestTree
+intGenTests = testGroup "Spec.Int" $
+    fromGroup <$> [
+      runGeneratorTestsWhere (Proxy :: Proxy Int) "Int Generator" (Yes :: Formula IntProp)
+    ]
```
We can automatically generate a test suite that checks that our parameterisedGenerator does what it is supposed to. This effectively gives us a way to verify that our model is sound by making sure that the model logic embedded in the generator matches the model logic embedded in the property definitions.

delicious-lemon commented on examples/Spec/Int.hs:
```diff
@@ -0,0 +1,89 @@
+intPureTests = testGroup "Pure.AcceptsSmallNegativeInts" $
+  fromGroup <$> [
+    runPureTestsWhere (Proxy :: Proxy Int) "AcceptsSmallNegativeInts" (Yes :: Formula IntProp)
+  ]
```
We can test a pure function using an instance of `HasPureTestRunner` like so.

delicious-lemon commented on examples/Spec/Int.hs:
```diff
@@ -0,0 +1,89 @@
+intPlutarchTests = testGroup "Plutarch.AcceptsSmallNegativeInts" $
+  fromGroup <$> [
+    runScriptTestsWhere (Proxy :: Proxy Int) (Proxy :: Proxy IntProp) "AcceptsSmallNegativeInts" Yes
+  ]
```
There is also a Plutus Script backend that can be used to test compiled Plutus scripts. We could define a model for `ScriptContext` in order to test a validator or minting policy. Here I have provided an example implementation in `Plutarch`.

delicious-lemon commented on examples/Spec/IntPermutationGen.hs:
```diff
@@ -0,0 +1,138 @@
+{-# LANGUAGE TypeFamilies #-}
+
+module Spec.IntPermutationGen (
```
This is the same model as in `Spec.Int` but I'm demonstrating a new feature introduced by this rewrite that makes construction of a `parameterisedGenerator` more straight-forward. In this simple example it ends up taking more lines of code to implement but it makes it much easier to extend and modify the generator when we extend or modify the model.

delicious-lemon commented on examples/Spec/IntPermutationGen.hs:
```diff
@@ -0,0 +1,138 @@
+  parameterisedGenerator = buildGen baseGen
+
+baseGen :: Gen Int
+baseGen = (Gen.int (linear minBound maxBound))
```
We define the generator in terms of a Strongly Connected Graph of generators that can permute/modify an existing model. Since this Graph is Strongly Connected we can generate an initial "base" model and then traverse the graph to generate a model that satisfies a given set of properties. We can plug this `PermutationGenerator` into the `HasParameterisedGenerator` class.

A `PermutationEdge` is a random function from model to model that satisfies a given contract. We can `match` on a model using a logical expression on the model properties and define a `contract` that the edge must obey in terms of a function from the old set of properties to the new set of properties.

These contracts are checked when the generator runs so they must be obeyed. We can generate a test suite for the edges that bootstraps the `PermutationGenerator` to test each edge's contract and this gives us feedback much like a type-checker does for our `PermutationEdge` random morphisms.

delicious-lemon commented on examples/Spec/IntPermutationGen.hs:
```diff
@@ -0,0 +1,138 @@
+    runScriptTestsWhere (Proxy :: Proxy Int) (Proxy :: Proxy IntProp) "AcceptsSmallNegativeInts" Yes
+  ]
+
+intPermutationGenSelfTests :: TestTree
+intPermutationGenSelfTests = testGroup "Int HasPermutationGenerator permutationGeneratorSelfTest" $
+  fromGroup <$> permutationGeneratorSelfTest (\(_ :: PermutationEdge Int IntProp) -> True) baseGen
```
This is the generated test suite that tests the contracts for each `PermutationEdge`.

delicious-lemon commented on examples/Spec/TicTacToe.hs:
```diff
@@ -0,0 +1,519 @@
+    | WinDeclared
+    deriving stock (Eq,Ord,Enum,Show,Bounded)
+
+instance LogicalModel TicTacToeProperty where
+  logic = All [Var FromBoardInInitialState :<->: (All $ Var <$> [FromBoardIsCorrectSize
+                                                                ,FromBoardIsEmpty
+                                                                ])
+              ,Var FromBoardHasOddNumberOfPieces :->: Not (Var FromBoardIsEmpty)
+              ,Var IsPlayersTurn :<->:
+                   ((Var PlayerIsX :&&: (Not $ Var FromBoardHasOddNumberOfPieces))
+                 :||: ((Not $ Var PlayerIsX) :&&: (Var FromBoardHasOddNumberOfPieces)))
+              , Var FromBoardIsEmpty :->: Var FromBoardHasEqualNumberOfPieces
+              , Var FromBoardHasEqualNumberOfPieces :->: (Not $ Var FromBoardHasOddNumberOfPieces)
+              , Var FromBoardHasOneMoreXThanO :->: Var FromBoardHasOddNumberOfPieces
+              , Var ToBoardEqualToFromBoard :->: (All [
+                   Var FromBoardIsCorrectSize :<->: Var ToBoardIsCorrectSize
+                 , Var FromBoardIsEmpty :<->: Var ToBoardIsEmpty
+                                                      ])
+              , (All $ Var <$> [FromBoardIsCorrectSize
+                               ,FromBoardIsEmpty
+                               ,ToBoardIsCorrectSize
+                               ,ToBoardIsEmpty]) :->: Var ToBoardEqualToFromBoard
+              ]
```
Here is an unfinished example of a more complex model with a generator written using `HasPermutationGenerator`. The `PermutationGenerator` self-test test suite that checks correctness of the `PermutationGenerator` morphisms runs around 1500 tests in 40 seconds which is fast enough to give compiler like feedback during development.

delicious-lemon commented on examples/Spec/IntPermutationGen.hs:
```diff
@@ -0,0 +1,138 @@
+    runPureTestsWhere (Proxy :: Proxy Int) "AcceptsSmallNegativeInts" (Yes :: Formula IntProp)
+                ]
+
+instance HasPlutusTestRunner Int IntProp where
+  expect _ _ = Var IsSmall :&&: Var IsNegative
+  script _ i =
+    let ii = (fromIntegral i) :: Integer
+     in compile (pif (((fromInteger ii) #< ((fromInteger 0) :: Term s PInteger)) #&& (((fromInteger (-10)) :: Term s PInteger) #<= (fromInteger ii))) (pcon PUnit) perror)
+
+intPermutationGenPlutarchTests :: TestTree
+intPermutationGenPlutarchTests = testGroup "Plutarch.AcceptsSmallNegativeInts" $
+  fromGroup <$> [
+    runScriptTestsWhere (Proxy :: Proxy Int) (Proxy :: Proxy IntProp) "AcceptsSmallNegativeInts" Yes
+  ]
```
When we hook up the target piece of code we are testing (in this case some `Plutarch`) we provide an expression in terms of the model properties that defines when we expect the code to VALIDATE/ERROR. We get a property test for every possible combination of properties under the model generated for us.

delicious-lemon commented on src/Proper/HasPermutationGenerator.hs:
```diff
@@ -0,0 +1,301 @@
+        mGen = buildGen bgen
+        graph = buildGraph pedges
+        isco = isStronglyConnected graph
+     in if isco
+           then case findDupEdgeNames of
+                  [] -> testEdge ns pedges mGen <$> filter pefilter generators
+                  dups -> [Group "HasPermutationGenerator edge names must be unique." $
+                           [(fromString $ dup <> " not unique", property $ failure)
+                           | dup <- dups]
+                          ]
+           else [Group "HasPermutationGenerator Graph Not Strongly Connected" $
+                          [(fromString "Not strongly connected", abortNotSCC ns graph)]
+                        ]
```
The `PermutationGenerator` self-test suite checks that the edge names are unique and the graph is strongly connected. If the graph is not strongly connected we get a hint as to which edges need to be defined.

delicious-lemon commented on src/Proper/HasPermutationGenerator.hs:
```diff
@@ -0,0 +1,301 @@
+
+  buildGen :: forall t . Monad t => Gen m -> Set p -> PropertyT t m
+  buildGen g = do
+    let pedges = findPermutationEdges (Proxy :: Proxy m) (Proxy :: Proxy p)
+        edges = Map.keys pedges
+        distmap = distanceMap edges
+        (sn,ns) = numberNodes (Proxy :: Proxy m) (Proxy :: Proxy p)
+        graph = buildGraph pedges
+        isco = isStronglyConnected graph
```
We compute the graph and distance matrix up front when we build the generator which allows us to traverse the graph over random shortest paths extremely quickly.

Geometer1729 commented on examples/Spec/TicTacToe.hs:
```diff
@@ -0,0 +1,519 @@
+
+genNot9 :: Int -> Int -> Gen Int
+genNot9 l u = do
+  b <- Gen.bool
```
Why not `Gen.choice [Gen.int (linear l 8),Gen.Int (linear 10 u)]` ?

Geometer1729 replied:
There are a couple other places with the Gen.bool if pattern instead of Gen.choice too.

delicious-lemon replied:
Cool! That's definitely an improvement :+1:

delicious-lemon committed:
refactor some generators

delicious-lemon committed:
monadic contract syntax as suggested by Geometer1729

delicious-lemon committed:
some work toward model embedding and composition

it is now possible to compose generators

next steps - composing property ASTs
               - deriving Enumerable
               - embedding logic with AST traversal
                 - related to https://github.com/mlabs-haskell/apropos-tx/issues/3

delicious-lemon committed:
Formula AST enables embedding with Functor

delicious-lemon committed:
add TH for generating enumerable

delicious-lemon committed:
applicative instance for Formula

delicious-lemon committed:
lifting PermutationEdge from submodel

delicious-lemon committed:
update readme

delicious-lemon committed:
fix up (Int,Int) model with Plutarch

delicious-lemon committed:
fix error in graph bounds on construction

delicious-lemon committed:
allow reuse of submodel genSatisfying in baseModel generator

delicious-lemon committed:
update IntPair model with new baseGen submodel reuse style

delicious-lemon committed:
some tictactoe modelling + improved error message

we now correctly detect when PermutationEdges are unnecessary

delicious-lemon committed:
simplify examples

we can find a better example than TicTacToe that's more complex than the
int examples
perhaps an auction contract or an oracle

delicious-lemon committed:
renaming - proposed name change brutus + brutus.plutus

we can split the repo in twain
  - brutus can have no Plutus dependencies
  - brutus-plutus can have the Brutus.Plutus part of the module tree

delicious-lemon committed:
remove Applicative Formula - Applicative is not a sensible thing to do

delicious-lemon committed:
not on liftEdges - perhaps we should use lens

delicious-lemon committed:
note on liftEdges

adamczykm committed:
Fix some typos in the README, make tests multithreaded.

adamczykm commented on README.md:
```diff
@@ -1,11 +1,165 @@
+It's a function.
+
+### And how do you write this logical model?
+Good question. First you start with a model. Then you describe it's properties.
```
There are some typos here and there in the README. I've fixed some of them in a small PR.

delicious-lemon replied:
Thanks!

adamczykm commented on README.md:
````diff
@@ -1,11 +1,165 @@
+This tells the test runner when to expect Passing/Failing of the device.
+
+```Haskell
+instance IsPlutusModel MyModel MyProperty where
````
I think the README is outdated. It should probably be `HasScriptRunner` instead of `IsPlutusModel`.  It is also included in my PR targeting this branch

adamczykm reviewed (changes requested):
Some initial thoughts:
- According to our style guide:
    > All module-level definitions, as well as `where`
    -binds, MUST have explicit type
    signatures. Type variables MUST have an explicit `forall`
     scoping them, and
    all type variables MUST have explicit kind signatures.
-  The code needs ‘linting’ and formatting
- There’s no docstring documentation, maybe some of the explanations you’ve written so far can be adapted
-  You can speed up the tests using : `ghc-options: -threaded -rtsopts -with-rtsopts=-N` (included in my pr)

delicious-lemon committed:
some models requre edge composition ++ should we reify Contract AST?

delicious-lemon committed:
remove experimental retry feature - it's not useful

delicious-lemon committed:
MoveSequence supports out of bounds game length

delicious-lemon committed:
rename MoveSequence to PlayerLocationSequencePair

delicious-lemon committed:
build MoveSequence gen from PlayerLocationSequencePair gen

delicious-lemon committed:
Merge branch 'adamczykm/fraser/generator-transformation-graph' into fraser/generator-transformation-graph

delicious-lemon committed:
quantifying over Script with parameterisedGenerator

delicious-lemon committed:
cost modelling

delicious-lemon committed:
some notes

delicious-lemon committed:
apropos-tx now has an Apropos.Tx (TODO)

delicious-lemon committed:
labeling the edges gives us histograms for the permutationGenerator

delicious-lemon committed:
fix show instance for Formula

delicious-lemon committed:
demonstrate recursive retry

delicious-lemon committed:
remove Debug.Trace

delicious-lemon commented:
> Some initial thoughts:
> 
>     * According to our style guide:
>       > All module-level definitions, as well as `where`
>       > -binds, MUST have explicit type
>       > signatures. Type variables MUST have an explicit `forall`
>       > scoping them, and
>       > all type variables MUST have explicit kind signatures.
> 
>     * The code needs ‘linting’ and formatting
> 
>     * There’s no docstring documentation, maybe some of the explanations you’ve written so far can be adapted
> 
>     * You can speed up the tests using : `ghc-options: -threaded -rtsopts -with-rtsopts=-N` (included in my pr)

I have merged this I think.

I'll get the linting and style sorted :+1:

delicious-lemon committed:
run fourmolu

delicious-lemon commented:
I would be in favor of merging this PR as is @adamczykm @Geometer1729. I have done some linting with fourmolu but I think it would be nice if we could separate the lint/formatting out into a separate PR - would be nice if we could just run `make format` for example.

delicious-lemon commented:
I think a separate PR that looks at the CI pipeline and formatting/linting would be good - it seems like we are compiling a lot of stuff in the pipeline that could be cached.

adamczykm commented:
I'm okay with doing linting and format in a separate PR. Let me take another look onto the code.
